cosmos/module.go

```diff
@@ -0,0 +1,118 @@
+package cosmos
+
+import (
+	"encoding/json"
+
+	"github.com/cosmos/cosmos-sdk/client/context"
+	"github.com/cosmos/cosmos-sdk/codec"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/types/module"
+	"github.com/gorilla/mux"
+	"github.com/spf13/cobra"
+	abci "github.com/tendermint/tendermint/abci/types"
+)
+
+// type check to ensure the interface is properly implemented
+var (
+	_ module.AppModule      = AppModule{}
+	_ module.AppModuleBasic = AppModuleBasic{}
+)
+
+// TODO: is it really needed to split AppModule and AppModuleBasic
+func NewAppModuleBasic(name string) AppModuleBasic {
+	return AppModuleBasic{
+		name: name,
+	}
+}
+
+func NewAppModule(moduleBasic AppModuleBasic, cdc *codec.Codec, handler sdk.Handler, querier Querier) AppModule {
+	return AppModule{
+		AppModuleBasic: moduleBasic,
+		handler:        handler,
+		querier:        querier,
+		cdc:            cdc,
+	}
+}
+
+type AppModuleBasic struct {
+	name string
+}
+
+func (m AppModuleBasic) Name() string {
+	return m.name
+}
+
+func (AppModuleBasic) RegisterCodec(cdc *codec.Codec) {
+}
+
+func (AppModuleBasic) DefaultGenesis() json.RawMessage {
+	return []byte("{}")
+}
+
+// ValidateGenesis checks the Genesis
+func (AppModuleBasic) ValidateGenesis(bz json.RawMessage) error {
+	return nil
+}
+
+// RegisterRESTRoutes registers rest routes
+func (AppModuleBasic) RegisterRESTRoutes(ctx context.CLIContext, rtr *mux.Router) {}
+
+// GetQueryCmd returns the root query command of this module
+func (AppModuleBasic) GetQueryCmd(cdc *codec.Codec) *cobra.Command {
+	return nil
+}
+
+// GetTxCmd returns the root tx command of this module
+func (AppModuleBasic) GetTxCmd(cdc *codec.Codec) *cobra.Command {
+	return nil
+}
+
+type Querier func(request sdk.Request, path []string, req abci.RequestQuery) (res interface{}, err error)
+
+type AppModule struct {
+	AppModuleBasic
+	handler sdk.Handler
+	querier Querier
+	cdc     *codec.Codec
+}
+
+func (m AppModule) RegisterInvariants(ir sdk.InvariantRegistry) {}
+
+func (m AppModule) Route() string {
+	return m.name
+}
+
+func (m AppModule) NewHandler() sdk.Handler {
+	return m.handler
+}
+func (m AppModule) QuerierRoute() string {
+	return m.name
+}
+
+func (m AppModule) NewQuerierHandler() sdk.Querier {
+	return func(request sdk.Request, path []string, req abci.RequestQuery) ([]byte, sdk.Error) {
+		data, err := m.querier(request, path, req)
+		if err != nil {
+			return nil, sdk.ErrInternal(err.Error())
+		}
```

Comment:
> I don't really like to force the error as an internal error. 

The think is cosmos expects that interface.

> We should just return the error that we have and m.querier should return a sdk.Error (not sure this is the best solution).

Querier is just helper struct that dosen't have to implement any interface thus it should stick with `error` type.

> The stuff I don't like here is that we will not have the possibility to have specific error like NotFound, AuthFailure or stuff like that that will provide good context.

I think we can. Internal is one of basic [codes](https://github.com/cosmos/cosmos-sdk/blob/master/types/errors.go#L30) and I guess the app can retrun any code type we want.

Reply:
> The stuff I don't like here is that we will not have the possibility to have specific error like `NotFound`, `AuthFailure` or stuff like that that will provide good context.

This can be improved over time. The goal here is to expose a simple API. We can imagine a system that cast error to cosmos error and return it if it is, otherwise create a generic error.